doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,191 @@
+
+Normalizing a `CREATE INDEX` statement will install the explicit `IN CLUSTER`
+clause if it was not explicitly specified.
+
+The syntax of `SHOW INDEXES` will be extended to allow limiting the indexes
+shown to one cluster:
+
+```
+SHOW { INDEX | INDEXES | KEYS } { FROM | IN } <on_name>
+[IN CLUSTER <cluster_name>]
+{ LIKE 'pattern | EXPR <expr> }
+```
+
+The output of `SHOW INDEXES` will be extended with an additional `cluster_name`
+column which displays the name of the cluster in which the index is
+materialized.
+
+### SQL session changes
+
+A new `cluster` session variable will be introduced that specifies the "active"
+cluster for a SQL session. It is analogous to the existing `database` session
+variable.
+
+The default value for the `cluster` session variable will be initially hardcoded
+to `default`. Users can change the value on a per-session basis via the usual
+mechanisms, like the pgwire startup protocol and the `SET` command.
+
+See [future work](#future-work) for details on how we plan to make the default
+configurable in the future.
+
+`SELECT` statements issued in a session will only use the indexes from the
+active cluster.
+
+### System catalog changes
+
+A new `mz_clusters` table with the following structure will describe the
+available clusters in the system:
+
+Field   | Type      | Meaning
+--------|-----------|--------
+id      | `bigint`  | The ID of the cluster.
+name    | `text`    | The name of the cluster.
+virtual | `boolean` | Whether the cluster is virtual.
+size    | `text`    | The size of the cluster.
+
+The existing `mz_indexes` table will be extended with a `cluster_id` field:
+
+Field      | Type      | Meaning
+-----------|-----------|--------
+...        | ...       | ...
+cluster_id | `bigint`  | The ID of the cluster in which the index is materialized.
+
+### Coordinator changes
+
+Internally, each cluster will be identified by an identifier in the range `[1,
+i64::MAX]`. The coordinator will be responsible for allocating identifiers and
+durably recording the mapping from names to identifiers via the following new
+table in the SQLite catalog:
+
+```sql
+CREATE TABLE clusters (
+    id   integer PRIMARY KEY,
+    name text NOT NULL UNIQUE
+);
+```
+
+This table works analogously to the existing `databases` table in the catalog.
+
+The migration that creates the table will additionally install a default row
+of `(1, 'default')`. This creates a cluster named `default` that is present by
+default in every Materialize installation. The default cluster is not otherwise
+special and can be dropped if desired.
+
+We will write a catalog migration that adds the `IN CLUSTER default` clause to
+all existing indexes.
+
+The various bookkeeping in the coordinator will need to change to account for
+clusters. Execution of a `SELECT` statement, for example, will need to select
+from only the indexes in the active cluster.
+
+TODO(mcsherry?): can we more specifically describe what bookkeeping will change?
```

Comment:
I think we can anticipate some bookkeeping changes, but I don't know I can confidently enumerate all of it. I'm happy to work with on this though. The `Controller` already maintains "which indices are available, with which `since` and `upper`, partitioned by compute instance", and we could teach `coord.rs` to rely on this, maybe (there are hiccups about transactions at the moment, with `coord.rs` holding back commands that prevent this state from updating).

`coord.rs` already has various uses of `DEFAULT_COMPUTE_INSTANCE` as an argument to `Controller` methods, and each of these would need to be populated. There are also some moment where `_instance` is ignored in `dataflow::Response` that `coord.rs` receives that could lead to confusion (the introspection indexes on each cluster are independent, and we will produce `upper` updates for each of them, and they conflict if conflated).